File: scripts/check_patch_review_cadence.mjs
#!/usr/bin/env node
// scripts/check_patch_review_cadence.mjs
/**
 * Registry review cadence monitor (non-blocking).
 *
 * Walks the two broken-EN retreat registries:
 *   - `EN_SOURCE_PATCHES` (segment-level)
 *   - `SOURCE_SYNC_EXCLUSIONS` (page-level, Phase A)
 *
 * Lists every entry whose `reviewAfter` date is in the past relative to `now`.
 * Warnings are printed to stderr (human consumable) and the process exits 0 —
 * this is monitoring, not a gate. CI may surface the warnings but must not
 * fail on them.
 *
 * Usage:
 *   node scripts/check_patch_review_cadence.mjs
 *   npm run check:patch-review
 *
 * Also re-exports the core logic (`collectOverduePatches`, `formatWarning`)
 * for reuse from inside `check_source_parity.mjs` (non-blocking warning
 * emission at run start) and from unit tests.
 *
 * Plan: docs/SYSTEM_SPEC.md §システム不変量
 */

import { EN_SOURCE_PATCHES } from './lib/en_source_patches.mjs';
import { SOURCE_SYNC_EXCLUSIONS } from './lib/source_sync_exclusions.mjs';

/**
 * Determine whether a patch's `reviewAfter` date has passed.
 *
 * @param {{reviewAfter: string}} patch
 * @param {number} nowMs - Current time in ms (injectable for tests).
 * @returns {{ overdue: boolean, daysOverdue: number, invalid: boolean }}
 */
export function evaluatePatchReview(patch, nowMs = Date.now()) {
  if (typeof patch?.reviewAfter !== 'string' || patch.reviewAfter.length === 0) {
    return { overdue: false, daysOverdue: 0, invalid: true };
  }
  const parsedMs = new Date(patch.reviewAfter).getTime();
  if (!Number.isFinite(parsedMs)) {
    return { overdue: false, daysOverdue: 0, invalid: true };
  }
  if (nowMs <= parsedMs) {
    return { overdue: false, daysOverdue: 0, invalid: false };
  }
  const daysOverdue = Math.floor((nowMs - parsedMs) / (24 * 60 * 60 * 1000));
  return { overdue: true, daysOverdue, invalid: false };
}

/**
 * Return every registry entry whose `reviewAfter` is in the past.
 *
 * @param {ReadonlyArray<object>} registry
 * @param {number} nowMs
 * @returns {Array<{ id: string, reviewAfter: string, daysOverdue: number }>}
 */
export function collectOverduePatches(registry = EN_SOURCE_PATCHES, nowMs = Date.now()) {
  const overdue = [];
  for (const patch of registry) {
    const result = evaluatePatchReview(patch, nowMs);
    if (result.overdue) {
      overdue.push({
        id: patch.id,
        reviewAfter: patch.reviewAfter,
        daysOverdue: result.daysOverdue,
      });
    }
  }
  return overdue;
}

/**
 * Same as `collectOverduePatches` but specialised for the page-level
 * `SOURCE_SYNC_EXCLUSIONS` registry (indexed by slug instead of id).
 *
 * @param {Record<string, object>} registry
 * @param {number} nowMs
 * @returns {Array<{ slug: string, reviewAfter: string, daysOverdue: number }>}
 */
export function collectOverdueSyncExclusions(
  registry = SOURCE_SYNC_EXCLUSIONS,
  nowMs = Date.now(),
) {
  const overdue = [];
  for (const [slug, entry] of Object.entries(registry)) {
    const result = evaluatePatchReview(entry, nowMs);
    if (result.overdue) {
      overdue.push({
        slug,
        reviewAfter: entry.reviewAfter,
        daysOverdue: result.daysOverdue,
      });
    }
  }
  return overdue;
}

/**
 * Render a single overdue-entry warning line.
 *
 * Accepts either an `en_source_patches` row (keyed by `id`) or a
 * `source_sync_exclusions` row (keyed by `slug`). The shape is detected
 * from which field is present.
 *
 * Precedence / validation:
 *   - `id` wins over `slug` (en_source_patches is the segment-level mechanism
 *     and its `id` uniquely names the patch; sync_exclusions has no `id`
 *     by construction, so the two shapes are disjoint in every current caller).
 *   - If neither `id` nor `slug` is set, an explicit "unknown-entry" label is
 *     emitted instead of silently stringifying `undefined`. This surfaces a
 *     programming error rather than hiding it in a log line.
 *
 * @param {{ id?: string, slug?: string, reviewAfter: string, daysOverdue: number }} entry
 * @returns {string}
 */
export function formatWarning(entry) {
  if (entry.id) {
    return `[en_source_patches] reviewAfter overdue: patch=${entry.id} reviewAfter=${entry.reviewAfter} daysOverdue=${entry.daysOverdue}`;
  }
  if (entry.slug) {
    return `[source_sync_exclusions] reviewAfter overdue: slug=${entry.slug} reviewAfter=${entry.reviewAfter} daysOverdue=${entry.daysOverdue}`;
  }
  return `[registry-review-cadence] reviewAfter overdue: entry=<unknown> reviewAfter=${entry.reviewAfter ?? '<none>'} daysOverdue=${entry.daysOverdue ?? 0}`;
}

/**
 * Main entry point used by CLI + npm script.
 *
 * @returns {{ overdueCount: number, exitCode: 0 }}
 */
export function main({
  patchRegistry = EN_SOURCE_PATCHES,
  exclusionsRegistry = SOURCE_SYNC_EXCLUSIONS,
  nowMs = Date.now(),
  stderr = console.warn,
  stdout = console.log,
} = {}) {
  const overduePatches = collectOverduePatches(patchRegistry, nowMs);
  const overdueExclusions = collectOverdueSyncExclusions(exclusionsRegistry, nowMs);
  const patchTotal = patchRegistry.length;
  const exclusionTotal = Object.keys(exclusionsRegistry).length;
  const overdueCount = overduePatches.length + overdueExclusions.length;

  if (overdueCount === 0) {
    stdout(
      `[registry-review-cadence] OK — ${patchTotal} en_source_patches + ` +
        `${exclusionTotal} source_sync_exclusions, 0 overdue`,
    );
    return { overdueCount: 0, exitCode: 0 };
  }
  stdout(
    `[registry-review-cadence] ${overdueCount} overdue entr(ies) ` +
      `(${overduePatches.length} patches / ${overdueExclusions.length} exclusions, ` +
      `warning only, non-blocking):`,
  );
  for (const entry of overduePatches) {
    stderr(formatWarning(entry));
  }
  for (const entry of overdueExclusions) {
    stderr(formatWarning(entry));
  }
  return { overdueCount, exitCode: 0 };
}

// CLI
if (import.meta.url === `file://${process.argv[1]}`) {
  const { exitCode } = main();
  process.exit(exitCode);
}

File: scripts/py/src/testim_parity/detection/check_patch_review_cadence.py
"""``scripts/detection/check_patch_review_cadence.mjs`` の Python port。

Registry review cadence monitor (non-blocking)。2 つの retreat registry
(``EN_SOURCE_PATCHES`` = segment-level / ``SOURCE_SYNC_EXCLUSIONS`` = page-level)
を走査し、``reviewAfter`` が過ぎている entry を stderr に warning 出力する。

Exit code は常に 0 — monitoring であって gate ではない。CI は warning を
surface できるが failed にしてはいけない。
"""

from __future__ import annotations

import sys
from collections.abc import Mapping, Sequence
from datetime import UTC, datetime
from typing import IO, Any

from ..en_source_patches import EN_SOURCE_PATCHES
from ..sync_exclusions import SOURCE_SYNC_EXCLUSIONS

__all__ = [
    "collect_overdue_patches",
    "collect_overdue_sync_exclusions",
    "evaluate_patch_review",
    "format_warning",
    "main",
]


def _now_ms() -> int:
    """現在時刻を UTC epoch ms で返す (mjs ``Date.now()`` 等価)。"""
    return int(datetime.now(tz=UTC).timestamp() * 1000)


def _iso_to_ms(date_str: str) -> float | None:
    """mjs ``new Date(str).getTime()`` 等価。invalid なら None。"""
    try:
        # YYYY-MM-DD は ISO 8601 として datetime.fromisoformat で parse 可能。
        # mjs ``new Date("2026-04-21")`` と同じく UTC midnight 扱い。
        dt = datetime.fromisoformat(date_str).replace(tzinfo=UTC)
    except ValueError:
        return None
    return dt.timestamp() * 1000


def evaluate_patch_review(
    patch: Mapping[str, Any] | None, now_ms: int | None = None
) -> dict[str, Any]:
    """patch の ``reviewAfter`` が過去なら overdue + days を返す (mjs 等価)。"""
    if now_ms is None:
        now_ms = _now_ms()
    review_after = (patch or {}).get("reviewAfter")
    if not isinstance(review_after, str) or len(review_after) == 0:
        return {"overdue": False, "daysOverdue": 0, "invalid": True}
    parsed_ms = _iso_to_ms(review_after)
    if parsed_ms is None:
        return {"overdue": False, "daysOverdue": 0, "invalid": True}
    if now_ms <= parsed_ms:
        return {"overdue": False, "daysOverdue": 0, "invalid": False}
    days_overdue = int((now_ms - parsed_ms) // (24 * 60 * 60 * 1000))
    return {"overdue": True, "daysOverdue": days_overdue, "invalid": False}


def collect_overdue_patches(
    registry: Sequence[Mapping[str, Any]] | None = None,
    now_ms: int | None = None,
) -> list[dict[str, Any]]:
    """``reviewAfter`` が過去の en_source_patches entry を集める (mjs 等価)。"""
    if registry is None:
        registry = EN_SOURCE_PATCHES
    if now_ms is None:
        now_ms = _now_ms()
    overdue: list[dict[str, Any]] = []
    for patch in registry:
        result = evaluate_patch_review(patch, now_ms)
        if result["overdue"]:
            overdue.append(
                {
                    "id": patch.get("id"),
                    "reviewAfter": patch.get("reviewAfter"),
                    "daysOverdue": result["daysOverdue"],
                }
            )
    return overdue


def collect_overdue_sync_exclusions(
    registry: Mapping[str, Mapping[str, Any]] | None = None,
    now_ms: int | None = None,
) -> list[dict[str, Any]]:
    """``reviewAfter`` が過去の source_sync_exclusions entry を集める (mjs 等価)。"""
    if registry is None:
        registry = SOURCE_SYNC_EXCLUSIONS
    if now_ms is None:
        now_ms = _now_ms()
    overdue: list[dict[str, Any]] = []
    for slug, entry in registry.items():
        result = evaluate_patch_review(entry, now_ms)
        if result["overdue"]:
            overdue.append(
                {
                    "slug": slug,
                    "reviewAfter": entry.get("reviewAfter"),
                    "daysOverdue": result["daysOverdue"],
                }
            )
    return overdue


def format_warning(entry: Mapping[str, Any]) -> str:
    """overdue entry の warning 文字列を生成 (mjs 等価、id > slug > unknown)。"""
    if entry.get("id"):
        return (
            f"[en_source_patches] reviewAfter overdue: patch={entry['id']} "
            f"reviewAfter={entry.get('reviewAfter')} daysOverdue={entry.get('daysOverdue')}"
        )
    if entry.get("slug"):
        return (
            f"[source_sync_exclusions] reviewAfter overdue: slug={entry['slug']} "
            f"reviewAfter={entry.get('reviewAfter')} daysOverdue={entry.get('daysOverdue')}"
        )
    return (
        f"[registry-review-cadence] reviewAfter overdue: entry=<unknown> "
        f"reviewAfter={entry.get('reviewAfter') or '<none>'} "
        f"daysOverdue={entry.get('daysOverdue') or 0}"
    )


def main(
    *,
    patch_registry: Sequence[Mapping[str, Any]] | None = None,
    exclusions_registry: Mapping[str, Mapping[str, Any]] | None = None,
    now_ms: int | None = None,
    stdout: IO[str] | None = None,
    stderr: IO[str] | None = None,
) -> dict[str, int]:
    """CLI エントリポイント (mjs 等価)。``{"overdueCount": int, "exitCode": 0}`` を返す。"""
    stdout = stdout or sys.stdout
    stderr = stderr or sys.stderr
    registry = patch_registry if patch_registry is not None else EN_SOURCE_PATCHES
    exclusions = exclusions_registry if exclusions_registry is not None else SOURCE_SYNC_EXCLUSIONS
    if now_ms is None:
        now_ms = _now_ms()

    overdue_patches = collect_overdue_patches(registry, now_ms)
    overdue_exclusions = collect_overdue_sync_exclusions(exclusions, now_ms)
    patch_total = len(registry)
    exclusion_total = len(exclusions)
    overdue_count = len(overdue_patches) + len(overdue_exclusions)

    if overdue_count == 0:
        print(
            f"[registry-review-cadence] OK — {patch_total} en_source_patches + "
            f"{exclusion_total} source_sync_exclusions, 0 overdue",
            file=stdout,
        )
        return {"overdueCount": 0, "exitCode": 0}
    print(
        f"[registry-review-cadence] {overdue_count} overdue entr(ies) "
        f"({len(overdue_patches)} patches / {len(overdue_exclusions)} exclusions, "
        f"warning only, non-blocking):",
        file=stdout,
    )
    for entry in overdue_patches:
        print(format_warning(entry), file=stderr)
    for entry in overdue_exclusions:
        print(format_warning(entry), file=stderr)
    return {"overdueCount": overdue_count, "exitCode": 0}


if __name__ == "__main__":
    result = main()
    sys.exit(int(result["exitCode"]))
